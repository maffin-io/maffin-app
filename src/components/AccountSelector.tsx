import React from 'react';
import { useHotkeys } from 'react-hotkeys-hook';
import Select, {
  SingleValue,
  components,
} from 'react-select';
import { BiSearch } from 'react-icons/bi';
import useSWRImmutable from 'swr/immutable';

import { Account } from '@/book/entities';
import { getAccountsWithPath } from '@/book/queries';

export type AccountSelectorProps = {
  id?: string,
  placeholder?: string,
  onChange?: Function,
  showRoot?: boolean,
  isClearable?: boolean,
  ignoreAccounts?: string[],
  defaultValue?: Account,
  className?: string,
};

export default function AccountSelector(
  {
    placeholder,
    onChange,
    ignoreAccounts,
    id = 'accountSelector',
    showRoot = false,
    isClearable = true,
    defaultValue,
    className = '',
  }: AccountSelectorProps,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  useHotkeys('meta+k', () => {
    ref.current?.focus();
  });

  let { data: accounts } = useSWRImmutable(
    '/api/accounts',
    getAccountsWithPath,
  );

  accounts = accounts || [];
  accounts = accounts.filter(account => !(ignoreAccounts || []).includes(account.type));
  if (!showRoot) {
    accounts = accounts.filter(account => account.type !== 'ROOT');
  }

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={accounts}
      placeholder={placeholder || 'Choose account'}
      onChange={(newValue: SingleValue<Account> | null) => {
        if (onChange) {
          onChange(newValue);
        }
      }}
      isClearable={isClearable}
      defaultValue={defaultValue}
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      getOptionLabel={(option) => option.path}
      getOptionValue={(option) => option.path}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: () => '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none',
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiSearch className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
