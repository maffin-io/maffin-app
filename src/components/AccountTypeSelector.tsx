import React from 'react';
import Select, {
  SingleValue,
  components,
} from 'react-select';
import { BiShuffle } from 'react-icons/bi';

import { Account } from '@/book/entities';

export type AccountTypeSelectorProps = {
  id?: string,
  placeholder?: string,
  onChange?: Function,
  disabled?: boolean,
  ignoreTypes?: string[],
  className?: string,
};

export default function AccountTypeSelector(
  {
    placeholder,
    onChange,
    ignoreTypes = [],
    id = 'accountSelector',
    disabled = false,
    className = '',
  }: AccountTypeSelectorProps,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  const types = Account.TYPES.slice(1).filter(
    type => !ignoreTypes.includes(type),
  ).map(type => ({ type }));

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={types}
      isDisabled={disabled}
      placeholder={placeholder || 'Choose account type'}
      onChange={(newValue: SingleValue<{ type: string }> | null) => {
        if (onChange) {
          onChange(newValue?.type);
        }
      }}
      isClearable
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      getOptionLabel={(option) => option.type}
      getOptionValue={(option) => option.type}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: ({ isDisabled }) => {
          if (isDisabled) {
            return '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-700 !border-none';
          }
          return '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none';
        },
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiShuffle className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
