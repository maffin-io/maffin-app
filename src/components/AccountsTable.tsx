import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';
import classNames from 'classnames';

import Table from '@/components/Table';
import type { AccountsTree } from '@/types/accounts';

export type AccountsTableProps = {
  tree: AccountsTree,
};

export default function AccountsTable(
  {
    tree,
  }: AccountsTableProps,
): JSX.Element {
  return (
    <Table<AccountsTree>
      columns={columns}
      data={tree.children}
      initialSort={{ id: 'total', desc: true }}
      showHeader={false}
      showPagination={false}
      tdClassName="p-2"
      getSubRows={row => row.children}
    />
  );
}

const columns: ColumnDef<AccountsTree>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-20" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-20" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-20" />
          </button>
        )}
        <Link
          className={classNames('badge hover:text-slate-300', {
            'bg-green-500/20 text-green-300': row.original.account.type === 'INCOME',
            'bg-red-500/20 text-red-300': row.original.account.type === 'EXPENSE',
            'bg-cyan-500/20 text-cyan-300': ['ASSET', 'BANK'].includes(row.original.account.type),
            'bg-orange-500/20 text-orange-300': row.original.account.type === 'LIABILITY',
            'bg-violet-500/20 text-violet-300': ['STOCK', 'MUTUAL'].includes(row.original.account.type),
          })}
          href={`/dashboard/accounts/${row.original.account.guid}`}
        >
          {row.original.account.name}
        </Link>
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTree) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];
