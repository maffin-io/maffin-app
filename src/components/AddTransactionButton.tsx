import React from 'react';
import { BiPlusCircle } from 'react-icons/bi';

import TransactionForm from '@/components/forms/transaction/TransactionForm';
import { Account } from '@/book/entities';
import Modal from '@/components/Modal';
import Tooltip from '@/components/Tooltip';

type AddTransactionButtonProps = {
  account: Account,
};

export default function AddTransactionButton(
  {
    account,
  }: AddTransactionButtonProps,
): JSX.Element {
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);
  const disabled = ['INCOME', 'EXPENSE'].includes(account.type);

  return (
    <>
      <Modal
        title={`${account.path} transaction`}
        open={isModalOpen}
        setOpen={setIsModalOpen}
      >
        <TransactionForm
          onSave={() => { setIsModalOpen(false); }}
          account={account}
        />
      </Modal>
      <Tooltip
        text="Add transactions from accounts that are not INCOME/EXPENSE/STOCK/MUTUAL"
        show={disabled}
      >
        <button
          type="button"
          className="btn-primary"
          disabled={disabled}
          onClick={() => setIsModalOpen(!isModalOpen)}
        >
          <BiPlusCircle className="mr-1" />
          Add Transaction
        </button>
      </Tooltip>
    </>
  );
}
