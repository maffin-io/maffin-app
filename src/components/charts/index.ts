export { default as Bar } from './Bar';
export { default as Line } from './Line';
export { default as Tree } from './Tree';
export { default as Pie } from './Pie';

export { default as NetWorthHistogram } from './NetWorthHistogram';
