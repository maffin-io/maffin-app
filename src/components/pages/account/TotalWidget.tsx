import React from 'react';

import { useAccountsTotals } from '@/hooks/api';
import Money from '@/book/Money';
import type { Account } from '@/book/entities';
import StatisticsWidget from '@/components/StatisticsWidget';
import TotalChange from '@/components/widgets/TotalChange';

export type TotalWidgetProps = {
  account: Account,
};

export default function TotalWidget({
  account,
}: TotalWidgetProps): JSX.Element {
  const { data: t0 } = useAccountsTotals();
  const total0 = t0?.[account.guid] || new Money(0, account.commodity.mnemonic);

  return (
    <StatisticsWidget
      className="mr-2"
      title="Total"
      stats={total0.format()}
      description={<TotalChange account={account} />}
    />
  );
}
