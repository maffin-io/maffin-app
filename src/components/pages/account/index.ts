export { default as TotalLineChart } from './TotalLineChart';
export { default as SplitsHistogram } from './SplitsHistogram';
export { default as TransactionsTable } from './TransactionsTable';
export { default as Header } from './Header';
export { default as AccountInfo } from './AccountInfo';
export { default as InvestmentInfo } from './InvestmentInfo';
export { default as AssetInfo } from './AssetInfo';
