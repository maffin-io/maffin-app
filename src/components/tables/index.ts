export { default as InvestmentsTable } from './InvestmentsTable';
export { default as AccountsTable } from './AccountsTable';
export { default as TransactionsTable } from './TransactionsTable';
export { default as PricesTable } from './PricesTable';
