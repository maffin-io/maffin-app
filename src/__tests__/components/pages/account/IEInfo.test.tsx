import React from 'react';
import { render } from '@testing-library/react';
import { DateTime, Interval } from 'luxon';

import { IEInfo } from '@/components/pages/account';
import TotalWidget from '@/components/pages/account/TotalWidget';
import type { Account } from '@/book/entities';
import { TotalsPie, MonthlyTotalHistogram } from '@/components/charts';
import { AccountsTable } from '@/components/tables';

jest.mock('@/components/charts/MonthlyTotalHistogram', () => jest.fn(
  () => <div data-testid="MonthlyTotalHistogram" />,
));

jest.mock('@/components/charts/TotalsPie', () => jest.fn(
  () => <div data-testid="TotalsPie" />,
));

jest.mock('@/components/tables/AccountsTable', () => jest.fn(
  () => <div data-testid="AccountsTable" />,
));

jest.mock('@/components/pages/account/TotalWidget', () => jest.fn(
  () => <div data-testid="TotalWidget" />,
));

jest.mock('@/components/tables/AccountsTable', () => jest.fn(
  () => <div data-testid="AccountsTable" />,
));

describe('IEInfo', () => {
  let account: Account;
  beforeEach(() => {
    jest.spyOn(DateTime, 'now').mockReturnValue(DateTime.fromISO('2023-03-01') as DateTime<true>);
    account = {
      guid: 'guid',
      name: 'Expenses',
      type: 'EXPENSE',
      commodity: {
        mnemonic: 'EUR',
      },
    } as Account;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected when not placeholder', () => {
    const { container } = render(
      <IEInfo account={account} />,
    );

    expect(MonthlyTotalHistogram).toBeCalledWith(
      {
        guids: [account.guid],
        title: '',
        interval: Interval.fromDateTimes(
          DateTime.now().minus({ year: 1 }).startOf('month'),
          DateTime.now(),
        ),
      },
      {},
    );
    expect(TotalWidget).toBeCalledWith(
      { account },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('renders as expected when placeholder', () => {
    account.placeholder = true;
    account.childrenIds = ['1', '2'];
    const { container } = render(
      <IEInfo account={account} />,
    );

    expect(AccountsTable).toBeCalledWith({ guids: ['1', '2'] }, {});
    expect(TotalsPie).toBeCalledWith(
      {
        guids: ['1', '2'],
        showDataLabels: false,
        showTooltip: true,
        title: 'Total spent',
      },
      {},
    );
    expect(MonthlyTotalHistogram).toBeCalledWith(
      {
        guids: ['1', '2'],
        title: '',
        interval: Interval.fromDateTimes(
          DateTime.now().minus({ year: 1 }).startOf('month'),
          DateTime.now(),
        ),
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });
});
