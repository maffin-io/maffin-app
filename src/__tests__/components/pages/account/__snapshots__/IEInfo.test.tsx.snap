// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IEInfo renders as expected when not placeholder 1`] = `
<div>
  <div
    class="grid md:grid-cols-12 items-start"
  >
    <div
      class="grid md:grid-cols-12 auto-cols-fr items-start col-span-6"
    >
      <div
        class="md:col-span-4"
      >
        <div
          data-testid="TotalWidget"
        />
      </div>
      <div
        class="card md:col-span-8"
      />
    </div>
    <div
      class="col-span-6"
    >
      <div
        class="card"
      >
        <div
          data-testid="MonthlyTotalHistogram"
        />
      </div>
      <div
        class="col-span-12"
      />
    </div>
  </div>
</div>
`;

exports[`IEInfo renders as expected when placeholder 1`] = `
<div>
  <div
    class="grid md:grid-cols-12 items-start"
  >
    <div
      class="grid md:grid-cols-12 auto-cols-fr items-start col-span-6"
    >
      <div
        class="card md:col-span-6"
      >
        <div
          data-testid="TotalsPie"
        />
        <div
          class="mt-4"
        >
          <div
            data-testid="AccountsTable"
          />
        </div>
      </div>
      <div
        class="card md:col-span-6"
      />
    </div>
    <div
      class="col-span-6"
    >
      <div
        class="card"
      >
        <div
          data-testid="MonthlyTotalHistogram"
        />
      </div>
      <div
        class="col-span-12"
      />
    </div>
  </div>
</div>
`;
